streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -425,6 +469,34 @@ object ZStream extends Serializable {
       } yield Control(pull, Command.noop)
     }
 
+  /**
+   * The stream that always halts with `cause`.
+   *
+   * @tparam The error type
+   * @param the cause for halting the stream
+   * @return a stream that is halted
+   */
+  def halt[E](cause: => Cause[E]): Stream[E, Nothing] =
+    fromEffect(ZIO.halt(cause))
+
+  /**
+   * The infinite stream of iterative function application: a, f(a), f(f(a)), f(f(f(a))), ...
+   */
+  def iterate[A](a: A)(f: A => A): UStream[A] =
+    ZStream {
+      Managed.effectTotal {
+        var currA = a
```

Comment:
Do we want to encourage this style in the ZStream module? It was previously confined to StreamEffect (for no real reason).

Reply:
Only perhaps for performance-sensitive combinators? of which this one is not, so I'll switch to a `Ref` instead.

Reply:
Well I'd prefer they all be implemented in the most performant way possible. We previously forwarded to `StreamEffect`'s companion object for constructors that don't involve effects but that isn't really necessary.

Reply:
`StreamEffect` was a nice place to encapsulate the throwing and catching of signal exceptions and keep it separate from the ZIO code in ZStream.

Reply:
Ok, I just saw that. I'm fine to revert if you want to.